File: configs/skills/telemetry/scaffolds/telemetry_sdk.py
"""OmniSight Telemetry SDK — Python client stub.

C17 L4-CORE-17 Telemetry backend.
"""

from __future__ import annotations

import gzip
import json
import logging
import signal
import threading
import time
import traceback
from collections import deque
from typing import Any, Optional

logger = logging.getLogger(__name__)


class TelemetryClient:
    def __init__(
        self,
        endpoint_url: str,
        device_id: str,
        profile: str = "default",
        opt_in: bool = False,
        batch_size: int = 50,
        flush_interval: float = 60.0,
        max_queue_size: int = 5000,
        offline_queue_enabled: bool = True,
        compression: str = "gzip",
    ):
        self.endpoint_url = endpoint_url
        self.device_id = device_id
        self.profile = profile
        self.opt_in = opt_in
        self.batch_size = batch_size
        self.flush_interval = flush_interval
        self.max_queue_size = max_queue_size
        self.offline_queue_enabled = offline_queue_enabled
        self.compression = compression

        self._queue: deque[dict[str, Any]] = deque(maxlen=max_queue_size)
        self._lock = threading.Lock()
        self._connected = True
        self._running = True

        self._flush_thread = threading.Thread(
            target=self._flush_loop, daemon=True
        )
        self._flush_thread.start()

        self._install_crash_handlers()

    def _install_crash_handlers(self) -> None:
        original_hook = getattr(signal, "getsignal", lambda _: None)
        try:
            signal.signal(signal.SIGSEGV, self._crash_handler)
            signal.signal(signal.SIGABRT, self._crash_handler)
        except (OSError, ValueError):
            pass

    def _crash_handler(self, sig: int, frame: Any) -> None:
        tb = traceback.format_stack(frame) if frame else []
        self.send_crash(
            signal_name=signal.Signals(sig).name,
            stack_trace="".join(tb),
        )
        self.flush()

    def send_crash(
        self,
        signal_name: str,
        stack_trace: str,
        extra: Optional[dict[str, Any]] = None,
    ) -> bool:
        event = {
            "event_type": "crash_dump",
            "device_id": self.device_id,
            "timestamp": time.time(),
            "crash_signal": signal_name,
            "stack_trace": stack_trace,
        }
        if extra:
            event.update(extra)
        return self._enqueue(event)

    def send_usage(
        self,
        event_name: str,
        metadata: Optional[dict[str, Any]] = None,
    ) -> bool:
        event = {
            "event_type": "usage_event",
            "device_id": self.device_id,
            "timestamp": time.time(),
            "event_name": event_name,
        }
        if metadata:
            event["metadata"] = metadata
        return self._enqueue(event)

    def send_metric(
        self,
        metric_name: str,
        value: float,
        unit: str = "",
        tags: Optional[dict[str, str]] = None,
    ) -> bool:
        event = {
            "event_type": "perf_metric",
            "device_id": self.device_id,
            "timestamp": time.time(),
            "metric_name": metric_name,
            "metric_value": value,
        }
        if unit:
            event["metric_unit"] = unit
        if tags:
            event["metric_tags"] = tags
        return self._enqueue(event)

    def _enqueue(self, event: dict[str, Any]) -> bool:
        if not self.opt_in:
            return False
        with self._lock:
            if len(self._queue) >= self.max_queue_size:
                return False
            self._queue.append(event)
        return True

    def flush(self) -> int:
        with self._lock:
            events = list(self._queue)
            self._queue.clear()

        if not events:
            return 0

        for i in range(0, len(events), self.batch_size):
            batch = events[i : i + self.batch_size]
            self._send_batch(batch)

        return len(events)

    def _send_batch(self, batch: list[dict[str, Any]]) -> bool:
        if not self._connected and not self.offline_queue_enabled:
            return False

        payload = json.dumps({
            "device_id": self.device_id,
            "events": batch,
            "opt_in": self.opt_in,
        }).encode()

        if self.compression == "gzip":
            payload = gzip.compress(payload)

        # Stub: POST payload to self.endpoint_url
        logger.debug("Would POST %d events (%d bytes) to %s",
                      len(batch), len(payload), self.endpoint_url)
        return True

    def _flush_loop(self) -> None:
        while self._running:
            time.sleep(self.flush_interval)
            try:
                self.flush()
            except Exception:
                logger.exception("Flush failed")

    def set_connected(self, connected: bool) -> None:
        was_disconnected = not self._connected
        self._connected = connected
        if connected and was_disconnected and self.offline_queue_enabled:
            self.flush()

    def shutdown(self) -> None:
        self._running = False
        self.flush()

    @property
    def queue_size(self) -> int:
        return len(self._queue)

File: configs/skills/telemetry/scaffolds/telemetry_sdk.c
/**
 * OmniSight Telemetry SDK — C implementation stub.
 * C17 L4-CORE-17 Telemetry backend.
 */

#include "telemetry_sdk.h"

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* --- Internal state --- */

typedef struct {
    char     event_type[32];
    char     payload[4096];
    uint64_t timestamp_ms;
} queued_event_t;

static struct {
    bool              initialized;
    bool              opted_in;
    bool              connected;
    telemetry_config_t config;
    queued_event_t    *queue;
    uint32_t           queue_head;
    uint32_t           queue_tail;
    uint32_t           queue_count;
} g_tel = {0};

/* --- Signal handler for crash capture --- */

static void _crash_signal_handler(int sig) {
    if (!g_tel.initialized) _exit(128 + sig);
    /* Capture minimal crash dump — real implementation would walk the stack */
    char buf[256];
    snprintf(buf, sizeof(buf), "{\"signal\":%d}", sig);
    telemetry_send_crash(sig == SIGSEGV ? "SIGSEGV" : "SIGABRT",
                         "(stack trace unavailable in stub)", buf);
    telemetry_flush();
    signal(sig, SIG_DFL);
    raise(sig);
}

/* --- Public API --- */

telemetry_status_t telemetry_init(const telemetry_config_t *cfg) {
    if (!cfg || !cfg->endpoint_url || !cfg->device_id)
        return TEL_ERR_INVALID;

    g_tel.config      = *cfg;
    g_tel.opted_in    = cfg->opt_in;
    g_tel.connected   = true;
    g_tel.queue_head  = 0;
    g_tel.queue_tail  = 0;
    g_tel.queue_count = 0;

    uint32_t qs = cfg->max_queue_size ? cfg->max_queue_size : 1000;
    g_tel.queue = (queued_event_t *)calloc(qs, sizeof(queued_event_t));
    if (!g_tel.queue) return TEL_ERR_INVALID;

    g_tel.initialized = true;

    /* Install crash handlers */
    signal(SIGSEGV, _crash_signal_handler);
    signal(SIGABRT, _crash_signal_handler);

    return TEL_OK;
}

telemetry_status_t telemetry_shutdown(void) {
    if (!g_tel.initialized) return TEL_ERR_NOT_INIT;
    telemetry_flush();
    free(g_tel.queue);
    g_tel.queue = NULL;
    g_tel.initialized = false;
    return TEL_OK;
}

telemetry_status_t telemetry_send_crash(const char *sig, const char *stack_trace,
                                         const char *extra_json) {
    if (!g_tel.initialized) return TEL_ERR_NOT_INIT;
    if (!g_tel.opted_in)    return TEL_ERR_NO_CONSENT;

    uint32_t max = g_tel.config.max_queue_size ? g_tel.config.max_queue_size : 1000;
    if (g_tel.queue_count >= max) return TEL_ERR_QUEUE_FULL;

    queued_event_t *ev = &g_tel.queue[g_tel.queue_tail % max];
    strncpy(ev->event_type, "crash_dump", sizeof(ev->event_type) - 1);
    snprintf(ev->payload, sizeof(ev->payload),
             "{\"crash_signal\":\"%s\",\"stack_trace\":\"%s\",\"extra\":%s}",
             sig ? sig : "UNKNOWN",
             stack_trace ? stack_trace : "",
             extra_json ? extra_json : "{}");

    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    ev->timestamp_ms = (uint64_t)ts.tv_sec * 1000 + ts.tv_nsec / 1000000;

    g_tel.queue_tail++;
    g_tel.queue_count++;
    return TEL_OK;
}

telemetry_status_t telemetry_send_usage(const char *event_name, const char *metadata_json) {
    if (!g_tel.initialized) return TEL_ERR_NOT_INIT;
    if (!g_tel.opted_in)    return TEL_ERR_NO_CONSENT;

    uint32_t max = g_tel.config.max_queue_size ? g_tel.config.max_queue_size : 1000;
    if (g_tel.queue_count >= max) return TEL_ERR_QUEUE_FULL;

    queued_event_t *ev = &g_tel.queue[g_tel.queue_tail % max];
    strncpy(ev->event_type, "usage_event", sizeof(ev->event_type) - 1);
    snprintf(ev->payload, sizeof(ev->payload),
             "{\"event_name\":\"%s\",\"metadata\":%s}",
             event_name ? event_name : "",
             metadata_json ? metadata_json : "{}");

    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    ev->timestamp_ms = (uint64_t)ts.tv_sec * 1000 + ts.tv_nsec / 1000000;

    g_tel.queue_tail++;
    g_tel.queue_count++;
    return TEL_OK;
}

telemetry_status_t telemetry_send_metric(const char *metric_name, double value,
                                          const char *unit) {
    if (!g_tel.initialized) return TEL_ERR_NOT_INIT;
    if (!g_tel.opted_in)    return TEL_ERR_NO_CONSENT;

    uint32_t max = g_tel.config.max_queue_size ? g_tel.config.max_queue_size : 1000;
    if (g_tel.queue_count >= max) return TEL_ERR_QUEUE_FULL;

    queued_event_t *ev = &g_tel.queue[g_tel.queue_tail % max];
    strncpy(ev->event_type, "perf_metric", sizeof(ev->event_type) - 1);
    snprintf(ev->payload, sizeof(ev->payload),
             "{\"metric_name\":\"%s\",\"metric_value\":%.6f,\"metric_unit\":\"%s\"}",
             metric_name ? metric_name : "",
             value,
             unit ? unit : "");

    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    ev->timestamp_ms = (uint64_t)ts.tv_sec * 1000 + ts.tv_nsec / 1000000;

    g_tel.queue_tail++;
    g_tel.queue_count++;
    return TEL_OK;
}

telemetry_status_t telemetry_flush(void) {
    if (!g_tel.initialized) return TEL_ERR_NOT_INIT;
    if (!g_tel.connected)   return TEL_ERR_NETWORK;

    /* Stub: in production, batch-POST queued events to endpoint_url */
    g_tel.queue_head  = g_tel.queue_tail;
    g_tel.queue_count = 0;
    return TEL_OK;
}

telemetry_status_t telemetry_set_consent(bool opted_in) {
    if (!g_tel.initialized) return TEL_ERR_NOT_INIT;
    g_tel.opted_in = opted_in;
    return TEL_OK;
}

uint32_t telemetry_queue_size(void) {
    return g_tel.queue_count;
}

bool telemetry_is_connected(void) {
    return g_tel.connected;
}
